### DISCUSSION TITLE
Multiple Jobs of Same Type running at same time when using DisallowConcurrentExecution

### DISCUSSION THREAD

**Discussion by mdmoura** [quartznet/quartznet#1565] [Q&A]
I need to run multiple `ManagerJob` each one using a different `Manager` configuration and same or different Cron:

```
public class ManagerJob : IJob {

  private readonly ILogger _logger;
  private readonly IManager _manager;

  public ManagerJob(IManager manager, ILogger logger) {
    _logger = logger;
    _manager = manager;
  } 
  
  public async Task Execute(IJobExecutionContext context) {

    _logger.Information($"Job Started @ { DateTime.Now }");

    await Task.Delay(TimeSpan.FromSeconds(20));

    await _manager.Run();

    _logger.Information($"Job Finished @ { DateTime.Now }");

  } 

}
```

I am configuring Quartz.Net on a Net 6 Worker project as:

```
  builder.ConfigureServices((context, services) => {

    services.AddQuartz(x => {

      x.SchedulerId = "Bot";

      x.UseMicrosoftDependencyInjectionJobFactory();
      x.UseJobFactory<JobFactory>();
        
      x.UseSimpleTypeLoader();
      x.UseInMemoryStore();
        
      x.UseDefaultThreadPool(y => { y.MaxConcurrency = 40; });

      foreach (Strategy strategy in strategies) {

        x.ScheduleJob<ManagerJob>(
          y => y.WithIdentity(strategy.Name).WithCronSchedule(strategy.Cron),
          y => y.WithIdentity(strategy.Name).UsingJobData(
              new JobDataMap((IDictionary<String, Object>) new Dictionary<String, Object> { { 
                 "Strategy", strategy 
               } }))
        );

      }

    });

    services.AddQuartzHostedService(x => {
      x.WaitForJobsToComplete = true;
    });

    services.AddTransient<IService, Service>();     

  });
```

I  implemented a `JobFactory` where I create an instance of `ManagerJob` with a `Manager` that uses a custom `Strategy`.

I am not injecting a `Manager` into `ManagerJob` because each one as a `Manager` with a different `Strategy`

```
public class JobFactory : IJobFactory {

  private readonly ILogger _logger;
  private readonly IService _service;
  
  public JobFactory(ILogger logger, IService service) {  
    _logger = logger;
    _service = service;
  } 
  
  public IJob NewJob(TriggerFiredBundle bundle, IScheduler scheduler) { 

    Strategy strategy = (Strategy)bundle.JobDetail.JobDataMap.Get("Strategy");

    IManager manager = new Manager(strategy, _logger, _service);
    
    return new ManagerJob(manager, _logger);
  
  } 

} 
```

I ran 2 Manager Jobs and they run at the same time as expected ...

But when I add the attribute `[DisallowConcurrentExecution]` they kept running at the same time.

Why?

**Side Questions:**

Is my approach the best one for this situation? 
Can I drop the `JobFactory `and create the Manager in `Execute` method of `ManagerJob`?
I am not sure this would be the way as a Manager would be created on each execution.
Is there any disadvantage of using a `JobFactory`? Maybe when using a persistent JobStore?

I feel like the documentation is not very clear how to run multiple Jobs of same type but different configuration.

**Comment by lahma** [comment]
You need to remember that `[DisallowConcurrentExecution]` locks based on job identity, so it applies to all job instances having same name and group. In your example you are creating multiple jobs with different identities. None of the should run concurrently with multiple triggers pointing to them, but they all can be running at se same time as long as they are different instances.

**Comment by mdmoura** [comment]
> You need to remember that [DisallowConcurrentExecution] locks based on job identity (...)

So if I give the same identity to all jobs and add the attribute `[DisallowConcurrentExecution]` no 2 jobs will run at the same time?

      foreach (Strategy strategy in strategies) {

        x.ScheduleJob<ManagerJob>(
          y => y.WithIdentity(strategy.Name).WithCronSchedule(strategy.Cron),
          y => y.WithIdentity("Manager").UsingJobData(
              new JobDataMap((IDictionary<String, Object>) new Dictionary<String, Object> { { 
                 "Strategy", strategy 
               } }))
        );

      }

In this situation the Jobs has the same identity (Manager) and the triggers different identities as `strategy.Name` is unique.

However with this configuration I get the error:

    'Quartz.SchedulerException' in System.Private.CoreLib.dll: 'Jobs added with no trigger must be durable.'

Why are jobs with No Trigger if I am adding both in ScheduleJob?

**Reply by lahma** [reply]
You can't give same identity to different jobs. Thinks job as database row, there can't be two rows with same unique key.

You should probably add the job separately *once* outside the loop and as durable and then the triggers you need and provide the strategy to triggers and reference the the single job key/instance for each trigger.
